### DISCUSSION TITLE
Why can't we have Collection Fixtures class-parallelable ?

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#3017] [Question] [answered]
👋🏻 Hi xUnit Team!

I'm trying to do some tricky stuff with xUnit + Collection Fixtures + databases using Testcontainers.

I'm trying to steal this idea of RavenDB -> they create a single instance of a DB _per "application" test run_ and then create a unique DB per test. I've actually got this working with TestContainers + MSSQL by leveraging ClassFixtures. Works great! Problem is ... stopping the DB container. I'm only having 1 instance of the container (ok ...) and then running all the tests in parallel (per class parallel). But if stop the container in the ClassFixture Dispose method, the other tests now can't complete.

When working with CollectionFixtures it works perfectly. Except ... the tests all run syncronously. WHICH TAKES FOREVER.

So here I am stuck between two different problems. Collection Fixtures were designed to have a _shared resource_ and because it's _shared_ then only one person can access this resource at a test-time. In my case, the shared resource is the MSSql _server isntance_ .. not the actual DB (tenant) in the server. Each test is getting their own DB (which means I need to create tables and seed data, etc. I'm happy with this). So I am getting isolation between tests still.

Maybe 2.x doesn't have any features for this .. but maybe 3.x does? I'm willing to jump on the -pre bandwagon to try and move on from setting up tests.

Last option here is the Assembly Fixture but i have no idea if that forces everything to run Sync also?

**Comment by bradwilson** [comment, marked as answer]
Assembly fixtures do not force everything to be sync.

Supporting more parallelization options: https://github.com/xunit/xunit/issues/2055 and https://github.com/xunit/xunit/issues/1986

**Reply by PureKrome** [reply]
🤗 Thank you Brad for replying so promptly. I had a quick play with v3 but had some migration issues. It's me - not you/v3. Right now I'm doing this:

```
// This runs Every Test (per class).
public abstract class BaseSqlServerTest(
    SqlServerFixture _sqlServerFixture,
    ITestOutputHelper _testOutputHelper,
    bool _isFakeDataSeeded = true) : IClassFixture<SqlServerFixture>, IAsyncLifetime
{
    public async Task InitializeAsync()
    {
        // Generate a unique database name using the test class name and the test name + post-pend GUID.
        // Can't have dashes. Needs to be < 128, so I max it at 100 chars.
       
        // Setup DB 
        // Create + seed (if requested to do that)

        // Create the Web Application Factory, etc.
    }

    public async Task DisposeAsync()
    {
        // Remove db connection strings, if not null, etc.
    }
}

// This runs ONCE per Class.
public class SqlServerFixture : IAsyncLifetime
{
    private readonly MsSqlContainer _msSqlContainer;

    public SqlServerFixture() =>
        _msSqlContainer = new MsSqlBuilder()
            .WithImage("mcr.microsoft.com/mssql/server:2022-CU13-ubuntu-22.04")
            .Build();

    public string ConnectionString => _msSqlContainer.GetConnectionString();

    public async Task InitializeAsync()
    {
        await _msSqlContainer.StartAsync();

        Console.WriteLine("** TestContainers: SQL Server is running.");
    }

    public async Task DisposeAsync()
    {
        if (_msSqlContainer != null)
        {
            await _msSqlContainer.StopAsync();
        }
    }
}
```

So it's not perfect.
- each class gets a container (this is the not-perfect bit)
- each test get's it's own DB (this is nice!)
- I am seeding every test. I flip this if i find that I don't want this every test.

So yeah .. i'll subscribe to those links and wait and see when things drop.

Love your work and keep fighting the good fight 💪🏻
